Enhancement restore external api restful interface

BenjamenMeyer opened the pull request:
This restore the RESTful API that was originally part of #647, and fixed per changes in related PRs that resulted in #647 getting merged.

This PR consists of two things:
1) The RESTful interface
2) The related tests

Tests are broken into separate files for each division of the interface (f.e OSKS-CATALOG vs OSKS-ADM). All tests pass.

Note: As this is only 3 commits (no I didn't squash anything), I hate the size but I have no good way of breaking this down in a good time constraint. I don't want to repeat the process of #647 and #648 that lead to this one - which will just be a messy merge/delete/commit/merge/delete commit.

BenjamenMeyer commented:
Cleaning up and removing tests that are nearly identical and DDT can provide the same coverage.

glyph commented:
@BenjamenMeyer - can you address the linting failure and merge master? Thanks!

BenjamenMeyer commented:
@glyph that's in progress. I've already got the linting fixed locally; getting a couple other things cleaned up then I'll merge master and push another commit set up. When done I'll remove the `work in progress` label and let you know.

glyph commented:
Sounds good!

BenjamenMeyer commented:
@glyph everything is ready to go - ready for review.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +675,518 @@ def list_tenants(self, request):
+
+        try:
+            service.update_template(endpoint_template_instance)
+        except ValueError:
```
I think I should have caught this in an earlier review, but `ValueError` and `IndexError` are bad exceptions to use here.  You can get an `IndexError` or a `ValueError` from basic type issues with built-in Python data structures, and so it's possible to mask basic implementation bugs here and make them very hard to diagnose.  Please fix `update_template` to raise a more specific `EndpointAlreadyExists`, `ServiceTypeMismatch` and `ServiceIdMismatch`.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +675,518 @@ def list_tenants(self, request):
+                    request
+                )
+            )
+        except IndexError:
```
Similarly, this should be `TemplateDoesntExist`

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +675,518 @@ def list_tenants(self, request):
+
+        try:
+            service = self.core.get_external_api(service_id)
+        except IndexError:
```
See point about generic exception types.

glyph commented on mimic/test/test_identity_osksadm.py:
```diff
@@ -0,0 +1,366 @@
+        api_list = []
+
+        # create the desired number of services per test parameter
+        for _ in range(api_entry_count):
```
I tend to prefer "ignored" for ignored loop variables; `_` is used as a special variable by a few other systems like `gettext` and the interactive interpreter.  No serious problem, but as a habit, it helps avoid potential clashes.

glyph replied:
Loops in tests are a code smell, and comprehensions can be nicer (when they work) for building up lists than manually calling `.append`.  Consider `api_list = [ExternalApiStore(...) for ignored in range(api_entry_count)]` instead.

glyph commented on mimic/test/test_identity_osksadm.py:
```diff
@@ -0,0 +1,366 @@
+        for _ in range(api_entry_count):
+            api_list.append(
+                ExternalApiStore(
+                    text_type(uuid.uuid4()),
```
`uuid.uuid4` actually does I/O (gathering entropy) so the values under test here won't be stable.  You might want to _not_ ignore the loop variable and use a `uuid5` instead, which will allow you to write this without copy/pasting big obnoxious lists but will still give you stable output.

BenjamenMeyer replied:
Actually, we really don't care about the stability of UUID4. It's perfectly okay for it to generate different values from one run to the next.

glyph replied:
If you don't want to change it in this case, that's fine; this is more general test advice.

glyph commented on mimic/test/test_identity_osksadm.py:
```diff
@@ -0,0 +1,366 @@
+            """
+            Lookup the API in the test's set of APIs  and match the values
+            """
+            for api in api_list:
```
Another way to write this, again, avoiding loops in tests:

```python
matching_apis = [api for api in api_list if api.uuid_key == api_id]
self.assertEqual(len(matching_apis), 1)
[matching_api] = matching_apis
# ... assert assert assert ...
```

glyph commented on mimic/test/test_identity_osksadm.py:
```diff
@@ -0,0 +1,366 @@
+        self.assertEqual(len(json_body["OS-KSADM:services"]), len(api_list))
+        # ensure all services in the response match one in the generated
+        # initially generated set
+        for entry in json_body["OS-KSADM:services"]:
```
Is it OK if this list is empty?

BenjamenMeyer replied:
Yes, it means no APIs have been registered. I believe that is part of the test via `ddt`; though can't remember if it's in this specific test or not.

glyph commented on mimic/test/test_identity_osksadm.py:
```diff
@@ -0,0 +1,366 @@
+        """
+        DELETE an unknown service will generate a 404.
+        """
+        data = {
```
There's a lot of code duplication between these tests. Could this be factored out into a common assertion?  It seems like you'd save a lot of lines that way.

glyph commented on mimic/test/test_identity_oskscatalog.py:
```diff
@@ -0,0 +1,688 @@
+        GET can retrieve numerous External APIs that have External API Templates.
+        """
+        api_list = [self.eeapi]
+        for _ in range(10):
```
Another place where a list comprehension would be a better fit, I think.  `api_list = [self.eeapi] + [ ... ]`

glyph commented:
My main complaint is still that this is a lot of code to review at once - but rather than splitting it up, I think just attacking the code duplication in the tests could get it down to a much more manageable size, and make the tests more maintainable in the future, to boot.  Can you give that a shot?  I don't see any blockers here though, aside from raw change size (I'm really trying to get individual merges down to <400 lines); I think if I can review this at a smaller size I'll be happy to push the green button.

BenjamenMeyer commented:
@glyph understand the size issue; I do try to remove duplication where I can - and `ddt` helps a lot with that. I'll see what I can do once I address the issues you list with the various `Exceptions`.  The tests do make up a significant portion of the code additions (typically 75% from what I noticed in previous PRs).

BenjamenMeyer commented:
I think I fixed the issues related to the Exceptions - I changed the ones in `core.py` and `model/identity_objects.py`. Still have other comments to address but wanted to get this up here.

BenjamenMeyer commented:
@glyph I've gone through and cleaned up some of the duplication using `ddt` and mix-ins; net of -207 lines in the last 3 commits.

glyph commented on mimic/core.py:
```diff
@@ -24,6 +24,66 @@
+    """
+    Parent for all Exceptions related to MimicCore
+    """
+    pass
```
The "pass" is unnecessary, since the docstring fills the syntactic niche that would require it to be here.  Can you remove these?

glyph commented on mimic/model/identity_errors.py:
```diff
@@ -0,0 +1,74 @@
+"""
+
+
+class EndpointTemplateExceptions(Exception):
```
Style nitpick; class names should be singular; except in rare cases where they represent a collection, and this one doesn't.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+
+        Reference: http://developer.openstack.org/api-ref-identity-v2-ext.html
+
+        Note: Either the service-id must be specified in the header or
```
Not worth changing it in this branch, but the correct way to format this with a ReST docstring is `.. note::`, as documented [here](http://sphinx.readthedocs.io/en/stable/markup/para.html?highlight=note#directive-note).

BenjamenMeyer commented:
@glyph I had to update against master after merging #666 so I went ahead and address the few comments I saw as well (between `pass`, plurality naming, and RST format).

glyph commented on mimic/model/identity_errors.py:
```diff
@@ -0,0 +1,74 @@
+    """
+    Parent for all Identity Endpoint Template Exceptions
+    """
+    pass
```
Can you remove all these `pass` statements?

BenjamenMeyer replied:
Fixed for next push.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+        List the available external services that endpoint templates
+        may be added to.
+
+        .. note:: Part of the OS-KSADM extension.
```
When adding a note like this, it's helpful to include a hyperlink that provides a direct reference so that future maintainers don't need to google around for a term.

BenjamenMeyer replied:
I've added URLs where I can still grab them...but it seems that OS-KSCATALOG is no longer being listed by OpenStack's documentation :frowning: so I used the URL I already had in there. It's the best I can do but also a reflection of the sad state of affairs regarding URLs and how easily they become non-existent.

glyph replied:
Yeah :-/.  Links going dead is just an unfortunate fact of life with this stuff.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+        may be added to.
+
+        .. note:: Part of the OS-KSADM extension.
+        .. note:: Does not implement the limits or markers.
```
Any time you have a reference in a docstring or a comment to desirable-but-unimplemented functionality, or a known bug, it should include a hyperlink referencing the place that the future work is tracked.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+        .. note:: Part of the OS-KSADM extension.
+        .. note:: Does not implement the limits or markers.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
This `x_auth_token` check ought to be a function, or better yet, a decorator that declares authorization is required.  Although mimic is not exactly security sensitive currently, duplicating this kind of code is a very risky anti-pattern; since mimic *may* eventually be handling administrative credentials (and this change only makes it more likely that it will be) we should treat it as though it were.  Please de-duplicate this with the other auth-token checks so that if it does more than check for the existence of the header in the future we don't have to re-implement it.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+            otherwise a UUID4 will be assigned to the 'id' field and the
+            'description' will be given a generic value.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+
+        .. note:: Part of the OS-KSADM extension.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+
+        .. note:: Marker/Limit capability not implemented here.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+            EndpointTemplateStore.required_mapping for details. Other
+            implementations may have different requirements.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+            a Service Name by the same name must already exist. Otherwise
+            a Not Found (404) will be returned.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+
+        .. note:: Marker/Limit capability not implemented here.
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+
+        Reference: http://developer.openstack.org/api-ref-identity-v2-ext.html
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented on mimic/rest/identity_api.py:
```diff
@@ -669,6 +688,521 @@ def list_tenants(self, request):
+
+        Reference: http://developer.openstack.org/api-ref-identity-v2-ext.html
+        """
+        x_auth_token = request.getHeader(b"x-auth-token")
```
Reference earlier comment `x-auth-token` duplication.

glyph commented:
I made a separate comment for each duplicated x-auth-token check which will hopefully make it easy to track the review response.

Beyond that duplication I think we are OK to land, so if you can address that, please feel free to push the green button when you're done.  However, in the future, let's please try to keep the changes smaller ;).  (For example, the exception-translation stuff in this branch could have been split out; but, at this point that would just be a finger exercise, since I've gone through and done the whole 2k-line review at this point...)

BenjamenMeyer commented:
@glyph please check the `rest/decorators.py` and `test/rest_decorators.py`. The rest should just be converting over to that. See https://github.com/rackerlabs/mimic/pull/665/commits/872dc382de8fcb9ae53ccfd3d5b1a70171e9f516 to keep it simple.
